### PR TITLE
Add CombinedListView

### PR DESCRIPTION
**Pull request by matanlurey**
Partially addresses https://github.com/dart-lang/core/issues/711

### CODE REVIEW

File: lib/src/combined_wrappers.dart [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import 'dart:collection';
+
+/// Returns a new list that represents [lists] flattened into a single list.
+///
+/// All methods and accessors treat the new list as-if it were a single
+/// concatenated list, but the underlying implementation is based on lazily
+/// accessing individual list instances.
+///
+/// The returned list is unmodifiable.
+List/*<T>*/ combineLists/*<T>*/(List<List/*<T>*/> lists) {
```
Comment (nex3) [inline lib/src/combined_wrappers.dart]: My intuition is that this should be exposed as a class, to better express that you're constructing a new view over objects rather than just concatenating lists. I'd also call the class `CombinedListView` to emphasize the view aspect further.

File: lib/src/combined_wrappers.dart [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+import 'dart:collection';
+
+/// Returns a new list that represents [lists] flattened into a single list.
+///
+/// All methods and accessors treat the new list as-if it were a single
+/// concatenated list, but the underlying implementation is based on lazily
+/// accessing individual list instances.
+///
+/// The returned list is unmodifiable.
```
Comment (nex3) [inline lib/src/combined_wrappers.dart]: This should be explicit that indexing and `length` are `O(lists)` rather than `O(1)`, and that the resulting list is unmodifiable.

File: lib/src/combined_wrappers.dart [hunk 3]
```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
```
Comment (nex3) [inline lib/src/combined_wrappers.dart]: I'd prefer this file just be `combined_list.dart`, possibly in a `combined_wrappers` directory.

File: lib/src/combined_wrappers.dart [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+/// All methods and accessors treat the new list as-if it were a single
+/// concatenated list, but the underlying implementation is based on lazily
+/// accessing individual list instances.
+///
+/// The returned list is unmodifiable.
+List/*<T>*/ combineLists/*<T>*/(List<List/*<T>*/> lists) {
+  return new _CombinedList/*<T>*/(lists);
+}
+
+class _CombinedList<T> extends ListBase<T> implements UnmodifiableListView<T> {
```
Comment (nex3) [inline lib/src/combined_wrappers.dart]: Why not `with UnmodifiableListMixin`?

File: lib/src/combined_wrappers.dart [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  static void _throw() {
+    throw new UnsupportedError('Cannot modify an unmodifiable List');
+  }
+
+  final List<List<T>> _lists;
+
+  _CombinedList(this._lists);
+
+  @override
```
Comment (nex3) [inline lib/src/combined_wrappers.dart]: This package doesn't use `@override`.

File: lib/src/combined_wrappers.dart [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  set length(int length) {
+    _throw();
+  }
+
+  @override
+  int get length => _lists.fold(0, (length, list) => length + list.length);
+
+  @override
+  T operator [](int index) {
+    final initialIndex = index;
```
Comment (nex3) [inline lib/src/combined_wrappers.dart]: `var`, here and elsewhere

File: lib/src/combined_wrappers.dart [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    for (var i = 0; i < _lists.length; i++) {
```
Comment (nex3) [inline lib/src/combined_wrappers.dart]: Why not `var list in _lists`?

File: lib/src/combined_wrappers.dart [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+      var list = _lists[i];
+      if (index < list.length) {
+        return list[index];
+      }
+      index -= list.length;
+    }
+    throw new RangeError.value(initialIndex, 'index', 'Length of $length');
```
Comment (nex3) [inline lib/src/combined_wrappers.dart]: This should probably be [`new RangeError.index()`](https://api.dartlang.org/stable/1.21.1/dart-core/RangeError/RangeError.index.html).

File: test/combined_wrappers_test.dart [hunk 9]
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) 2013, the Dart project authors.  Please see the AUTHORS file
```
Comment (nex3) [inline test/combined_wrappers_test.dart]: 2017

File: test/combined_wrappers_test.dart [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+void main() {
+  final list1 = const [1, 2, 3];
+  final list2 = const [4, 5, 6];
+  final list3 = const [7, 8, 9];
+  final concat = []..addAll(list1)..addAll(list2)..addAll(list3);
+  final combine = combineLists([list1, list2, list3]);
+
+  // In every way possible this should test the same as an UnmodifiableListView.
+  common.testUnmodifiableList(concat, combine, 'combineLists');
```
Comment (nex3) [inline test/combined_wrappers_test.dart]: Also test an empty source list and a source list containing empty lists.

**Review by nex3 — CHANGES_REQUESTED** [review]

File: lib/src/combined_wrappers.dart (hunk 1 above)
Comment (matanlurey) [inline lib/src/combined_wrappers.dart]: Isn't the fact that it is a `CombinedListView` an implementation detail?

From a maintenance perspective you're adding an interface that others could implement/extend or reference even though there isn't any actual real use for doing so.

File: lib/src/combined_wrappers.dart (hunk 4 above)
Comment (matanlurey) [inline lib/src/combined_wrappers.dart]: `UnmodifiableListMixin` is part of the SDK and not public :-/

https://github.com/dart-lang/collection/issues/7

File: lib/src/combined_wrappers.dart (hunk 7 above)
Comment (matanlurey) [inline lib/src/combined_wrappers.dart]: Mostly dart2js-related paranoia (no reason to allocate an iterator)

File: test/combined_wrappers_test.dart (hunk 10 above)
Comment (matanlurey) [inline test/combined_wrappers_test.dart]: Done.

File: test/combined_wrappers_test.dart (hunk 9 above)
Comment (matanlurey) [inline test/combined_wrappers_test.dart]: Done.

File: lib/src/combined_wrappers.dart (hunk 8 above)
Comment (matanlurey) [inline lib/src/combined_wrappers.dart]: Done.

File: lib/src/combined_wrappers.dart (hunk 6 above)
Comment (matanlurey) [inline lib/src/combined_wrappers.dart]: Done.

File: lib/src/combined_wrappers.dart (hunk 5 above)
Comment (matanlurey) [inline lib/src/combined_wrappers.dart]: Done.

File: lib/src/combined_wrappers.dart (hunk 2 above)
Comment (matanlurey) [inline lib/src/combined_wrappers.dart]: Done.

File: lib/src/combined_wrappers.dart (hunk 3 above)
Comment (matanlurey) [inline lib/src/combined_wrappers.dart]: Done.

File: lib/src/combined_wrappers/combined_list.dart [hunk 11]
```diff
@@ -10,8 +10,13 @@ import 'dart:collection';
 /// concatenated list, but the underlying implementation is based on lazily
 /// accessing individual list instances.
 ///
-/// The returned list is unmodifiable.
+/// The resulting list has an index operator (`[]`) and `length` property that
+/// are both `O(lists)`, rather than `O(1)`, and the list is unmodifiable.
 List/*<T>*/ combineLists/*<T>*/(List<List/*<T>*/> lists) {
+  // Small optimization when there are no lists to avoid allocation.
+  if (lists.isEmpty) {
```
Comment (nex3) [inline lib/src/combined_wrappers/combined_list.dart]: Maybe `lists.every((list) => list.isEmpty)`?

File: lib/src/combined_wrappers.dart (hunk 1 above)
Comment (nex3) [inline lib/src/combined_wrappers.dart]: > Isn't the fact that it is a `CombinedListView` an implementation detail?

I think a class better expresses what's actually happening. Functions generally produce outputs that are, after the function is finished running, largely independent of their arguments. Classes are much more likely to retain a reference to their constructor argument and update their state based on changes, and the word "View" explicitly indicates the way that will happen.

It's also more consistent with similar existing patterns—we have `Unmodifiable*View` classes, not `make*Unmodifiable()` functions.

> I could have it return an `UnmodifiableListView<T>` though.

*Hopefully* no one is relying on that static type anywhere.

File: lib/src/combined_wrappers/combined_list.dart (hunk 11 above)
Comment (matanlurey) [inline lib/src/combined_wrappers/combined_list.dart]: Done, though without the closure allocation.

Also added an optimization for a single list, which is probably common enough too.

File: test/combined_wrappers_test.dart (hunk 10 above)
Comment (nex3) [inline test/combined_wrappers_test.dart]: I'd like to also test a source list containing empty lists, but not *only* empty lists.

**Review by nex3 — CHANGES_REQUESTED** [review]

File: lib/src/combined_wrappers/combined_list.dart [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+/// All methods and accessors treat the new list as-if it were a single
+/// concatenated list, but the underlying implementation is based on lazily
+/// accessing individual list instances.
+///
+/// The resulting list has an index operator (`[]`) and `length` property that
+/// are both `O(lists)`, rather than `O(1)`, and the list is unmodifiable.
+List/*<T>*/ combineLists/*<T>*/(List<List/*<T>*/> lists) {
+  // Small optimization when there are no lists to avoid allocation.
+  if (lists.isEmpty) {
+    return const [];
```
Comment (matanlurey) [inline lib/src/combined_wrappers/combined_list.dart]: Hmm. I agree with this. Changed it up so the underlying list can always be changed.

File: lib/src/combined_wrappers/combined_list.dart [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+import 'dart:collection';
+
+/// Returns a new list that represents [lists] flattened into a single list.
+///
+/// All methods and accessors treat the new list as-if it were a single
+/// concatenated list, but the underlying implementation is based on lazily
+/// accessing individual list instances.
+///
+/// The resulting list has an index operator (`[]`) and `length` property that
+/// are both `O(lists)`, rather than `O(1)`, and the list is unmodifiable.
```
Comment (matanlurey) [inline lib/src/combined_wrappers/combined_list.dart]: Done.

**Comment by matanlurey** [comment]
Friendly ping @nex3

**Comment by nex3** [comment]
Oh, I didn't realize this was ready for review—[this comment](https://github.com/dart-lang/collection/pull/50#discussion_r98311210) is still outstanding.

File: lib/src/combined_wrappers.dart (hunk 1 above)
Comment (matanlurey) [inline lib/src/combined_wrappers.dart]: Alright. I still don't really agree with this change, but willing to overlook it.

PTAL.

File: lib/src/combined_wrappers/combined_list.dart [hunk 14]
```diff
@@ -23,17 +23,17 @@ List/*<T>*/ combineLists/*<T>*/(List<List/*<T>*/> lists) {
   if (lists.length == 1) {
     return new UnmodifiableListView/*<T>*/(lists.first);
   }
-  return new _CombinedList/*<T>*/(lists);
+  return new CombinedList/*<T>*/._(lists);
 }
 
-class _CombinedList<T> extends ListBase<T> implements UnmodifiableListView<T> {
+class CombinedList<T> extends ListBase<T> implements UnmodifiableListView<T> {
```
Comment (nex3) [inline lib/src/combined_wrappers/combined_list.dart]: `CombinedListView`

Also, remove `combinedList()` in favor of a `CombinedListView` constructor.

File: lib/src/combined_wrappers/combined_list.dart [hunk 15]
```diff
@@ -10,63 +10,66 @@ import 'dart:collection';
...
-/// The returned list is unmodifiable.
+/// The resulting list has an index operator (`[]`) and `length` property that
+/// are both `O(lists)`, rather than `O(1)`, and the list is unmodifiable - but
+/// underlying changes to these lists are still accessible from the resulting
+/// list.
 List/*<T>*/ combineLists/*<T>*/(List<List/*<T>*/> lists) {
-  return new _CombinedList/*<T>*/(lists);
+  // Small optimization when there are no lists to avoid allocation.
+  if (lists.isEmpty) {
+    return const [];
```
Comment (nex3) [inline lib/src/combined_wrappers/combined_list.dart]: This still behaves inconsistently when `lists` itself is modified. I think we should just construct a `CombinedList()` regardless of the size of the input, and allow `lists` to change after the fact just like its contents.

**Review by nex3 — CHANGES_REQUESTED** [review]

File: lib/src/combined_wrappers/combined_list.dart (hunk 14 above)
Comment (matanlurey) [inline lib/src/combined_wrappers/combined_list.dart]: Done.

File: lib/src/combined_wrappers/combined_list.dart (hunk 15 above)
Comment (matanlurey) [inline lib/src/combined_wrappers/combined_list.dart]: Done.

**Review by nex3 — COMMENTED** [review]
Please add a test for modifying the `List<List>`—then this'll be good to go!

**Comment by matanlurey** [comment]
@nex3 Added the test case!
